controller/system_restore_controller.go

```diff
@@ -0,0 +1,397 @@
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *SystemRestoreController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Info("Starting Longhorn SystemRestore controller")
+	defer c.logger.Info("Shutting down Longhorn SystemRestore controller")
+
+	if !cache.WaitForNamedCacheSync(c.name, stopCh, c.cacheSyncs...) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *SystemRestoreController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *SystemRestoreController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncSystemRestore(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *SystemRestoreController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	log := c.logger.WithField("systemRestore", key)
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		log.WithError(err).Warn("Error syncing Longhorn SystemRestore")
+
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	log.WithError(err).Warn("Dropping Longhorn SystemRestore out of the queue", key)
+	c.queue.Forget(key)
+}
+
+func (c *SystemRestoreController) getLoggerForSystemRestore(name string) *logrus.Entry {
+	return c.logger.WithField("systemRestore", name)
+}
+
+func (c *SystemRestoreController) syncSystemRestore(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync SystemRestore %v", c.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	if namespace != c.namespace {
+		return nil
+	}
+
+	backupTargetClient, err := getBackupTargetClientDefault(c.ds)
+	if err != nil {
+		return err
+	}
+
+	return c.reconcile(name, backupTargetClient)
+}
+
+func (c *SystemRestoreController) reconcile(name string, backupTargetClient engineapi.BackupTargetClientSystemBackupInterface) (err error) {
+	systemRestore, err := c.ds.GetSystemRestore(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	log := c.getLoggerForSystemRestore(systemRestore.Name)
+
+	if !c.isResponsibleFor(systemRestore) {
+		return nil
+	}
+
+	if systemRestore.Status.OwnerID != c.controllerID {
+		systemRestore.Status.OwnerID = c.controllerID
+		systemRestore, err = c.ds.UpdateSystemRestoreStatus(systemRestore)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Infof("Picked up by SystemRestore Controller %v", c.controllerID)
+	}
+
+	if !systemRestore.DeletionTimestamp.IsZero() {
```

Comment:
So if the SystemRestore gets deleted whenever it's in-progress or not, the corresponding job will be just deleted? If yes, what's the status of the system? Is it possible to recover?

Or users just need to restart a new SystemRestore to try again, but before it succeeds, the system is broken.

Reply:
> So if the SystemRestore gets deleted whenever it's in-progress or not, the corresponding job will be just deleted? 

Yes

> If yes, what's the status of the system?

Depending on the phase of the system restore/rollout. When in the progress of CR restoration, then the Deployment, DaemonSets are likely already applied/created. However, if in the midway of volume restoration, then the volume is likely still capable to finish the restoration, but the PV and PVC might not make it.

> 
> Or users just need to restart a new SystemRestore to try again, but before it succeeds, the system is broken.

Yes, the user can try again with another SystemRestore.

Reply:
Need to document in the doc.

Reply:
https://github.com/longhorn/website/pull/621#issuecomment-1340732989